---
title: "rasp: analysis of alternative splicing through splicing ratios"
author:
- name: Jinxing Wei
affiliation:
  - &uab Department of Mathematics, Autonomous University of Barcelona (UAB)
- name: Pol CuscÃ³
  affiliation:
  - &isglobal Bioinformatics Research Group in Epidemiolgy (BRGE), Barcelona Insitute for Global Health (ISGlobal)
- name: Juan R. Gonzalez
  affiliation:
  - &isglobal Bioinformatics Research Group in Epidemiolgy (BRGE), Barcelona Insitute for Global Health (ISGlobal)
  - &uab Department of Mathematics, Autonomous University of Barcelona (UAB)
  email: juanr.gonzalez@isglobal.org
date: "`r Sys.Date()`"
package: "`r pkg_ver('rasp')`"
output:
  BiocStyle::html_document:
    number_sections: true
    toc: yes
    fig_caption: yes
    toc_float: true
bibliography: rasp.bib
vignette: >
  %\VignetteIndexEntry{rasp: analysis of alternative splicing through splicing ratios}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---


```{r style, echo=FALSE, message=FALSE, warning=FALSE}
library("knitr")
opts_chunk$set(message = FALSE, error = TRUE, 
               warning = FALSE, cache=TRUE, comment = "")
```

# Getting started

This document gives an overview of the R package `rasp`, which provides statistical procedures to study differential splicing among conditions in the context of annotated genomes. That is, isoform information is summarized using the available annotation (at exon or transcript level) using any of the existing tools. This information can provided in RPKM as for instance [FluxCapacitor](http://big.crg.cat/services/flux\_capacitor) does, or using read counts on exons. The statistical methods used in `rasp` are described in @GonCalSam12. The method is based on comparing the variability of the relative expression among conditions using the methodology described in @Anderson06 that can be seen as an ANOVA for the case of having multivariate data (multiple isoforms or exons per gene). The current vesion of the package also allows the user to compare splicing patterns for more than two conditions and adjust for other covariates.

We should start the R session by loading the library as usual:


```{r, load_rasp}
library(rasp)
```

We will start loading into the workspace the data corresponding to the initial table of transcipt expression estimates provided by  @GonCalSam12. The RNA-seq data correspond to lymphoblastoid cell lines derived from 69 unrelated Nigerian individuals produced by @PicMarPai10. More than half billion 35-bp-long reads were sequenced. Reads were mapped to the human genome version hg19 using GEM mapper @GEM. After that, mapped reads were used to obtain transcipt expression estimates using GENCODE version 3c produced in the framework of the ENCODE project as the reference genome annotation. Finally, Flux Capacitor was used to produce estimates of transcript abundances measured as RPKMs. Data was filtered to have only information about protein-coding genes expressed in all individuals, having at least two annotated splice forms and isoforms with an expresion level of at least 1 RPKM in at least one individual. These filters lead to a total of 1654 genes and 4668 associated transcripts @GonCalSam12. The object containing this information can be loaded by typing:

```{r load_YRI}
data(YRI)
dim(YRI)
```


The object also contains information about gender of each sample that will be used as a grouping factor to illustrate comparison among conditions:

```{r get_group}
genderYRI <- attr(YRI, "gender")
head(genderYRI)
table(genderYRI)
```

We are aware that this is not the proper way of encapsulating omic data since, for instance, Bioconductor has better data infrastructures such as `RangedSummarizedExperiment` that are specifically designed to deal with this type of data. Nontheless, we allow the function pass the data in that format to facilitate the task for those users who are not familiar with Bioconductor or that want to analyze a subset of data obtained from public repositories. [Section Data in `RangedSummarizedExperiment` format] illustrate an example having this type of data.



# Data Visualization

For each gene $n$ alternative splice forms (transcripts or exons) are observed. Therfore, for a given gene, and for each individual $j$, the relative expression of each transcipt, $i$ $i=1, \ldots, n$ is computed as:
%
\begin{equation}
f_{ij} =  \frac{x_{ij}}{\sum_{i=1}^{n}x_{ij}} 
\end{equation}
%
where $x_{ij}$ denotes the transcript expression estimates (either as RPKMs or counts). Therefore, we consider as a outcome the vector of splicing proportions for the $j$ individual, $f_{j}=(f_{1j}, \ldots,f_{nj})$. This multivariate nature of the data is the reason why @GonCalSam12 proposed to use the methodology described in @Anderson06 to compare groups. 

This type of data is also known as compositional data and can be represented in the simplex using the function `plotTernary()` as illustrate Figure \ref{fig:plotTernary} that can be obtained by typing

```{r plotTernary, fig.cap="Ternary plot. Ternary plot corresponding to the splicing ratios of three transcripts (1st, 2nd and 3rd) belonging to gene ENSG00000160741."}
plotTernary(YRI, "ENSG00000160741", transcripts=1:3)
```


Notice that the argument `transcript` indicates the three transcripts to be represented in the Ternary plot. An obvious limitation of the Ternary Plot is that it can oly plot data for three different transcripts. In order to plot the splicing ratios for any number of transcripts one can use the function `{\tt `plotAllIso()`. The result is a plot showing the splicing ratios for all selected transcripts and individuals. 

```{r plotAllIso, fig.cap="Isforoms plot. Plot showing splicing ratios of all transcripts in gene ENSG00000005448."}
plotAllIso("ENSG00000005448", data = YRI, group = genderYRI,
                      inds = 1:10)
```
# Data analysis

A single gene can be analyzed using the `testRasp()` function. For instance, let us investigate whether the splicing ratios of the gene ENSG00000160741 containing 3 different transcripts are different between males and females (Figure \ref{fig:plotTernary}). 

The p-value can be obtained by typing

```{r test_gene} 
gene <- YRI[YRI$gene_id=="ENSG00000160741", ]
mod <- rasp(x=gene, group=genderYRI, 
            geneidCol = 1, expressionCols = 3:ncol(gene))
mod
```

The first of the two p-values, {\tt p.homogeneity}, represents the p-value resulting from the test for dispersion while the second one, {\tt pvalue}, is the resulting p-value from the location test. The location test is sensitive to hetereogeneity in the dispersion of points of each group and, as a result, both p-values should always be taken into account. In this case, the dispersion p-value is highly non-significant, while the location p-value shows that there are significant differences in the splicing patterns among males and females. The ternary plot in Figure \ref{fig_simplex_gender} also shows some differences across sexes. One can observe that females tend to have more RPKMs in the transcript ENST0000046.

\begin{figure}[!h]
\begin{center}
<<fig=TRUE, echo=FALSE>>=
plotTernary(YRI, "ENSG00000160741", transcripts=1:3, group=genderYRI)
@
\end{center}
\caption{Ternary plot corresponding to the splicing ratios of three transcripts (1st, 2nd and 3rd) belonging to gene ENSG00000160741 for males and females.}
\label{fig_simplex_gender}
\end{figure}

\clearpage

\newpage

This process can be performed using multiple processors when analyzing several genes. This procedure is implemented in the main function of \Rasp~ called \Rfunction{rasp()}:

<<>>=
ngenes <- 5
sel.genes <-  names(table(YRI[,1]))[1:ngenes]
res <- rasp(YRI[YRI[,1]%in%sel.genes, ], genderYRI, mc.cores = 1, 
            geneidCol = 1, expressionCols = 3:ncol(YRI))
@

Note that we have only selected 5 genes among the existing ones in order to avoid computational problems.  Also notice that the function has several arguments beside the data and group objects. Firstly, it allows the user to use several cores during the computations ({\tt mc.cores} argument). Secondly, three diffent tests are allowed. Finally, three arguments must be specified to tell which columns of the original dataset are the \emph{gene id}, \emph{transcript id} and \emph{expression} respectively.

The centroids of each group, the p-values and BH adjusted p-values can be obtained by executing:

<<>>=
print(res)
@

\subsection{Test for groups with more than two levels}

As shown previously {\tt rasp} performs a test that looks for differences in splicing ratios across all possible conditions of a certaing group variable. Sometimes researchers might be interested in testing whether the individuals of a specific subgroup or condition have significant differences in the splicing ratios compared to the rest of conditions. We have developed a test that, for a given group with more than two levels, performs a permutation test comparing each of the conditions against the rest. This can be done specifying {\tt testGroup = TRUE} in both the {\tt rasp} and {\tt testRasp} functions, as shown in the following code.

<<results=hide>>=
set.seed(1234)
g <- as.character(genderYRI)
g[g == "female"] <- ifelse(runif(sum(g == "female"))<.5, "female1", "female2")
resG <- rasp(YRI[YRI[,1]%in%sel.genes, ], factor(g), mc.cores = 1,
             geneidCol = 1, expressionCols = 3:ncol(YRI), testGroup = TRUE)
@ 

<<>>=
print(resG)
@ 

# References

\clearpage

\section{Session info}

<<>>=
sessionInfo()
@


